Add application form for individuals

chrismedrela opened the pull request:
This PR is **not** intended to be merged -- it's only to make leaving feedback easier.

gvwilson commented on workshops/admin.py:
```diff
@@ -11,11 +11,16 @@
 
+
+class RoleAdmin(admin.ModelAdmin):
+    list_display = ('name', 'verbose_name', 'show_in_training_request_form')
```
Why `verbose_name`? I.e., what requirement does this satisfy?

chrismedrela replied:
`verbose_name` is a name that is displayed in the application form in `Previous involvement with Software Carpentry or Data Carpentry` question. `name` is internal name that we cannot change.

For most roles, `verbose_name` and `name` are the same, but that's not always the case, i.e. I've introduced `Role(name='contributor', verbose_name='Contributed to lesson materials')`.

gvwilson commented on workshops/forms.py:
```diff
@@ -878,3 +878,51 @@ class Meta:
+
+class TrainingRequestForm(forms.ModelForm):
+    intro = HTML("""
+        <p>Any text at the beginning of the form?</p>
```
We run two-day instructor training classes for people who wish to learn how to teach Software Carpentry and Data Carpentry workshops. If you wish to apply to attend an upcoming workshop, please fill in the form below. We will notify applicants as spaces become available.  If you have any questions, please mail training@software-carpentry.org.

Please note that as a condition of taking this training:

1. You are required to abide by our code of conduct, which can be found at http://software-carpentry.org/conduct/.

2. You must complete three short tasks after the course in order to complete certification. The tasks are described at http://swcarpentry.github.io/instructor-training/checkout/, and take a total of approximately 2 hours.

3. You are expected to teach at a Software Carpentry or Data Carpentry workshop within 12 months of the course.

For more information on Software and Data Carpentry instructor training, please see http://swcarpentry.github.io/instructor-training/.

chrismedrela replied:
Done.

gvwilson commented on workshops/management/commands/fake_database.py:
```diff
@@ -39,10 +39,18 @@ def fake_airports(self, faker, count=5):
-        for role in roles:
-            Role.objects.create(name=role)
+        roles = [
+            ('debriefed', 'Debriefed', False),
```
Why is "Debriefed" a role?  Also, I don't understand the doc string on this method.

chrismedrela replied:
I don't know. This role was introduced along with all other roles here https://github.com/swcarpentry/amy/commit/2942f0ed9c8f192811eaf2025d0c5d2cdf3d745b

gvwilson replied:
I think we should take it out - @pbanaszkiewicz ?

gvwilson commented on workshops/models.py:
```diff
@@ -1424,3 +1429,193 @@ def long_status(self):
+class TrainingRequest(models.Model):
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    name = models.CharField(
```
Why not personal name and family name (same as Person)?

chrismedrela commented:
I deleted `debriefed` role (however, I didn't add a migration that deletes this role from database, is it ok?) and split `name` field into `personal` and `family`.

gvwilson commented:
Please add the migration that deletes the Debriefed role as well - let's 
scrape it clean.
Cheers,
Greg

chrismedrela commented:
@gvwilson, done.

chrismedrela commented:
@gvwilson, I'd like to add some tests before merging this one.

@pbanaszkiewicz, @gvwilson, There are some tasks with `role=debriefed` in production database. Migration no 0087 (the one that deletes `debriefed` role) aborts if there is any such task, so first we need to delete them.

chrismedrela commented:
Added tests.

gvwilson commented:
Please have the migration delete those task entries before deleting the 
role.

chrismedrela commented:
@gvwilson, done.

gvwilson commented:
Thanks - @pbanaszkiewicz, are you happy with the testing?

gvwilson commented:
Just ran it locally:
1. Looking good.
2. Please use a pull-down for selecting country.
3. Nothing showed up for the captcha at the bottom when running on my laptop; the form won't be accepted without the captcha. Is there a way to configure it for local testing?

chrismedrela commented:
ad 2. done

ad 3. I have the same issue on my machine. @pbanaszkiewicz?

pbanaszkiewicz commented:
@chrismedrela @gvwilson unfortunately that's limitation of Google reCaptcha - you had to have priv/pub keypair locally (envvars `AMY_RECAPTCHA_PUBLIC_KEY` and `AMY_RECAPTCHA_PRIVATE_KEY`) in order for captcha field to show up. I didn't check if that behavior changed.

gvwilson commented:
I have exported the variables, and the CAPTCHA now shows, but it keeps telling me the values I've entered are wrong. Here's the latest attempt:
![screen shot 2016-05-28 at 4 10 28 pm](https://cloud.githubusercontent.com/assets/911566/15629734/d2bc113c-24ee-11e6-9f2d-af623ab8c0d9.png)
I'm pretty sure I've got the right values...

pbanaszkiewicz commented:
@gvwilson okay I forgot to add. When `DEBUG is True`, the only value that works in captcha field is "PASSED".

gvwilson commented:
Yup, that's a big one :-)

gvwilson commented:
@chrismedrela What is the `show_in_training_request_form` field of `Role` for?

gvwilson commented:
Works for me once the env vars are set

chrismedrela commented:
@gvwilson, it's a way to control whether a role is an option under "Previous involvement with Software Carpentry or Data Carpentry" question.

gvwilson commented:
Are any roles *not* shown right now?

chrismedrela commented:
@gvwilson, only "tutor".

gvwilson commented:
What the hell is "tutor"? (checks database) OK, please delete that role, and change the three Task objects that use it to have a role of instructor instead of tutor.

chrismedrela commented:
@gvwilson, do you want to delete `show_in_training_request_form` field and show all roles in the form?

gvwilson commented:
Also, please change the "thank" you page to say:

~~~
Thank you for applying for instructor training. We will contact you shortly.
~~~

chrismedrela commented:
I'm working at the list and detailed view of TrainingRequests.

Here is how the list view looks like:
![image](https://cloud.githubusercontent.com/assets/1645996/15630152/0360dd14-252f-11e6-8d78-185bf987487a.png)
@gvwilson, 1) What columns do you want to include in this view? 2) Do you want filtering? If yes, filtering by which fields?

Here is detailed view:
![image](https://cloud.githubusercontent.com/assets/1645996/15630208/38b853a6-2530-11e6-8350-898f91ce6604.png)
Please let me know what should I change.

gvwilson commented:
This is plenty good for now - please include when you merge (since it'll help with testing). Good work!

pbanaszkiewicz commented on workshops/forms.py:
```diff
@@ -878,3 +878,76 @@ class Meta:
+        <p>For more information on Software and Data Carpentry instructor training,
+        please see <a href="http://swcarpentry.github.io/instructor-training/">
+        http://swcarpentry.github.io/instructor-training/</a>.</p>
+    """)
```
This huge chunk should end up in a template, not in the form.

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -1085,10 +1085,15 @@ def __str__(self):
+                                    null=False, blank=True, default='')
+    show_in_training_request_form = models.BooleanField(
+        default=False,
+        verbose_name='include in training request form?')
```
What is the purpose of this field?

chrismedrela replied:
It's a way to control whether a role is an option under "Previous involvement with Software Carpentry or Data Carpentry" question. There are two roles that are not included in the form: `debriefed` and `tutor`, both roles will be deleted.

gvwilson replied:
@chrismedrela and I have discussed this, and I've agreed that the two 
roles are to be deleted.

pbanaszkiewicz replied:
No worries, I was only asking to understand the changes better.

Question: we use a different approach to mark badges that are for instructors. We use `BadgeQuerySet.instructor_badges` -- why don't you want to hardcode roles to be included in the form this way?

chrismedrela replied:
I've just deleted `show_in_training_request_form` role, so no need for hardcoding anything.

gvwilson replied:
I think we should add a field to Badge to show whether or not it's an 
instructor badge - or maybe add tags to badges (though that seems like 
overkill, since the only special flavor right now is instructor).

chrismedrela commented:
@gvwilson, @pbanaszkiewicz, I've made all changes that you suggested.

chrismedrela commented:
@gvwilson, @pbanaszkiewicz, ready to be reviewed/merged.

gvwilson commented:
:+1:

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -1424,3 +1426,197 @@ def long_status(self):
+        verbose_name='What is your current occupation/career stage?',
+        help_text='Please choose the one that best describes you.',
+        null=False, blank=True, default='undisclosed',
+    )
```
```python
occupation = models.CharField(…, choices=ProfileUpdateRequest.OCCUPATION_CHOICES)
```

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -1424,3 +1426,197 @@ def long_status(self):
+    )
+    gender = models.CharField(
+        max_length=1,
+        choices=GENDER_CHOICES,
```
```python
gender = models.CharField(…, choices=ProfileUpdateRequest.GENDER_CHOICES)
```

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -1424,3 +1426,197 @@ def long_status(self):
+        choices=GENDER_CHOICES,
+        null=False, blank=False, default=Person.UNDISCLOSED,
+    )
+    gender_other = models.CharField(
```
Don't you have `build_choice_field_with_other_option` for this purpose?

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -1424,3 +1426,197 @@ def long_status(self):
+    teaching_frequency_expectation, teaching_frequency_expectation_other = build_choice_field_with_other_option(
+        choices=TEACHING_FREQUENCY_EXPECTATION_CHOICES,
+        verbose_name='How often would you expect to teach classes on Software'
+                     'or Data Carpentry Workshops after this training?',
```
Space ` ` missing between "Software" and "or".

pbanaszkiewicz commented on workshops/templates/forms/trainingrequest.html:
```diff
@@ -0,0 +1,38 @@
+{% load selectable_tags %}
+
+{% block content %}
+<p>We run two-day instructor training classes for people who wish to learn how to
```
I'd add SWC and DC logos here.

chrismedrela replied:
Do you mean sth like that?
![image](https://cloud.githubusercontent.com/assets/1645996/15635677/d501cc6e-25e5-11e6-8c75-014d2b433a7c.png)

pbanaszkiewicz commented on workshops/templates/forms/trainingrequest_confirm.html:
```diff
@@ -0,0 +1,9 @@
+{% load crispy_forms_tags %}
+{% load selectable_tags %}
+
+{% block content %}
```
If the logos are on the form page, they should be here too.

pbanaszkiewicz commented on workshops/models.py:
```diff
@@ -1424,3 +1426,197 @@ def long_status(self):
+    )
+    gender_other = models.CharField(
+        max_length=STR_LONG,
+        verbose_name=' ',
```
Why do other fields have `verbose_name` and you set it to empty here? It's inconsistent.

chrismedrela replied:
It's left empty, because it looks good in the form. See the example of `gender`:
![image](https://cloud.githubusercontent.com/assets/1645996/15635642/ff55d5b0-25e4-11e6-9d0c-98f68547d2b8.png)

In the case of areas of expertise, there is no "Other (enter below)" option, so it could be not clear what the text field is for if we left the label empty.

pbanaszkiewicz commented on workshops/templates/workshops/trainingrequest.html:
```diff
@@ -0,0 +1,58 @@
+  <tr><th>programming language usage frequency</th>
+      <td>{{ req.get_programming_language_usage_frequency_display }}</td></tr>
+  <tr><th>Why do you want to attend this training course?</th>
+      <td>{{ req.reason|linebreaks }}</td></tr>
```
Use `<pre>` instead of linebreaks.

pbanaszkiewicz commented on workshops/templates/workshops/trainingrequest.html:
```diff
@@ -0,0 +1,58 @@
+          {% empty %}
+            <li>No knowledge domain.</li>
+          {% endfor %}
+          </ul></td></tr>
```
Other areas of expertise not shown here.

pbanaszkiewicz commented on workshops/templates/workshops/trainingrequest.html:
```diff
@@ -0,0 +1,58 @@
+            <li>No previous involvement.</li>
+          {% endfor %}
+          </ul></td></tr>
+  <tr><th>previous experience</th>
```
What about previous training in teaching?

pbanaszkiewicz commented:
@chrismedrela @gvwilson All is looking good. I left some minor comments – ping me back once you address them.

chrismedrela commented:
@pbanaszkiewicz, ping.

pbanaszkiewicz commented:
+1 from me to merge. @gvwilson?

pbanaszkiewicz commented:
Actually hold on. @chrismedrela @gvwilson do we want support for languages people can teach with? Ie. should we implement #798 here once merged?

gvwilson commented:
Yes, we want to show what languages they can teach with - once we've merged #798, we can come back and extend this one.

pbanaszkiewicz commented:
This should be merged first since there's still some work to be done in #798.

@gvwilson if you don't have any further comments I'm going to merge and add issue re. languages people teach with.

gvwilson commented:
Woo hoo!
G
